bundles/org.openhab.binding.shieldtv/src/main/java/org/openhab/binding/shieldtv/internal/ShieldTVHandler.java

```diff
@@ -0,0 +1,790 @@
+import org.openhab.binding.shieldtv.internal.protocol.shieldtv.ShieldTVMessageParserCallbacks;
+import org.openhab.binding.shieldtv.internal.protocol.shieldtv.ShieldTVRequest;
+import org.openhab.core.OpenHAB;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.ThingStatusInfo;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ShieldTVHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * Significant portions reused from Lutron binding with permission from Bob A.
+ *
+ * @author Ben Rosenblum - Initial contribution
+ */
+@NonNullByDefault
+public class ShieldTVHandler extends BaseThingHandler implements ShieldTVMessageParserCallbacks {
+
+    private static final int DEFAULT_RECONNECT_MINUTES = 5;
+    private static final int DEFAULT_HEARTBEAT_SECONDS = 5;
+    private static final long KEEPALIVE_TIMEOUT_SECONDS = 30;
+
+    private static final String STATUS_INITIALIZING = "Initializing";
+
+    private final Logger logger = LoggerFactory.getLogger(ShieldTVHandler.class);
+
+    private @Nullable ShieldTVConfiguration config;
+    private int reconnectInterval;
+    private int heartbeatInterval;
+    private int sendDelay;
+
+    private @NonNullByDefault({}) SSLSocketFactory sslsocketfactory;
+    private @Nullable SSLSocket sslsocket;
+    private @Nullable BufferedWriter writer;
+    private @Nullable BufferedReader reader;
+
+    private @NonNullByDefault({}) ShieldTVMessageParser shieldtvMessageParser;
+
+    private final BlockingQueue<ShieldTVCommand> sendQueue = new LinkedBlockingQueue<>();
+
+    private @Nullable Future<?> asyncInitializeTask;
+
+    private @Nullable Thread senderThread;
+    private @Nullable Thread readerThread;
+
+    private @Nullable ScheduledFuture<?> keepAliveJob;
+    private @Nullable ScheduledFuture<?> keepAliveReconnectJob;
+    private @Nullable ScheduledFuture<?> connectRetryJob;
+    private final Object keepAliveReconnectLock = new Object();
+
+    private StringBuffer sbReader = new StringBuffer();
+    private String lastMsg = "";
+    private String thisMsg = "";
+    private int inMessage = 0;
+
+    private boolean isLoggedIn = false;
+    private String hostName = "";
+    private String currentApp = "";
+
+    private ShieldTVPKI shieldtvPKI = new ShieldTVPKI();
+
+    public ShieldTVHandler(Thing thing) {
+        super(thing);
+        shieldtvMessageParser = new ShieldTVMessageParser(this);
+    }
+
+    public void setHostName(String hostName) {
+        this.hostName = hostName;
+        updateState(CHANNEL_DEVICENAME, new StringType(hostName));
```

Comment:
Since it is just static metadata, device name should be a Thing Property instead of stored in a channel that will never change.

```suggestion
        thing.setProperty("devicename", hostname);
```

Reply:
I had created this as a channel with the intention of having other channels (eventually) be presented to the user with other info (app for example) as it was uncovered in the protocol.   Unfortunately, decoding the protocol has been slow going.  This also could potentially change if the user makes a change on the device.

Reply:
Thing Properties are presented to the user on the thing config page. They can be updated dynamically. The hostname is not going to change unless the user re-configures the device  and is not useful in rules so putting it in a channel seems like the wrong approach.

![image](https://user-images.githubusercontent.com/26532542/208722625-74b35981-3496-432f-aab7-e93f0389ae67.png)

Reply:
Sounds good.  I'll make the change.

Reply:
Done and committed.

Reply:
I think you need to add it to the thing definition xml as well:
![image](https://user-images.githubusercontent.com/26532542/208735272-411655d1-07a5-4419-8e11-13049792d760.png)

Reply:
You are correct.  Done and committed.